jormungandr/src/blockchain/storage.rs

```diff
@@ -180,105 +60,109 @@ enum StreamingError {
+                    .map(|block| Block::deserialize(block.as_ref()))
+                    .collect::<Result<Vec<_>, _>>()
+                    .map_err(Error::Deserialize)
+            })
     }
 
-    pub async fn put_block(&self, block: Block) -> Result<(), Error> {
-        self.run(move |connection| match connection.put_block(&block) {
-            Err(StorageError::BlockNotFound) => unreachable!(),
-            Err(e) => Err(e),
-            Ok(()) => Ok(()),
-        })
-        .await
+    pub fn put_block(&self, block: &Block) -> Result<(), Error> {
+        let id = block
+            .header
+            .hash()
+            .serialize_as_vec()
+            .map_err(Error::Serialize)?;
+        let parent_id = block
+            .header
+            .block_parent_hash()
+            .serialize_as_vec()
+            .map_err(Error::Serialize)?;
+        let chain_length = block.header.chain_length().into();
+        let block_info = BlockInfo::new(id, parent_id, chain_length);
+        self.storage
+            .put_block(
+                &block.serialize_as_vec().map_err(Error::Serialize)?[..],
+                block_info,
+            )
+            .map_err(Into::into)
     }
 
     /// Return values:
     /// - `Ok(stream)` - `from` is ancestor of `to`, returns blocks between them
     /// - `Err(CannotIterate)` - `from` is not ancestor of `to`
     /// - `Err(BlockNotFound)` - `from` or `to` was not found
     /// - `Err(_)` - some other storage error
-    pub async fn stream_from_to(
+    pub fn stream_from_to(
         &self,
-        from: HeaderHash,
-        to: HeaderHash,
+        from: &HeaderHash,
+        to: &HeaderHash,
     ) -> Result<impl Stream<Item = Result<Block, intercom::Error>>, Error> {
-        let (rh, rf) = intercom::stream_reply(BLOCK_STREAM_BUFFER_SIZE, self.logger.clone());
-        let iter = self
-            .run(
-                move |connection| match connection.is_ancestor(&from, &to)? {
-                    Some(distance) => {
-                        let to_info = connection.get_block_info(&to)?;
-                        Ok(BlockIterState::new(to_info, distance, identity))
-                    }
-                    None => Err(Error::CannotIterate),
-                },
-            )
-            .await?;
-        let pump = BlockSinkPump::start(iter, self.pool.clone(), rh);
-        let stream = rf.await.expect("unexpected channel error");
-        Ok(PumpedStream::new(
-            stream,
-            stream::unfold(pump, |pump| {
-                pump.pump().map(|res| {
-                    res.expect("unexpected channel error")
-                        .map(|pump| ((), pump))
-                })
-            }),
-        ))
+        let distance = self
+            .storage
+            .is_ancestor(from.as_bytes(), to.as_bytes())?
+            .ok_or(Error::CannotIterate)?;
+
+        let stream = futures::stream::iter(self.storage.iter(to.as_bytes(), distance)?)
+            .fuse()
```

Comment:
Why do you need to fuse it? The function signature only promises an `impl Stream`.